### PR TITLE
WIP: added functional connectivity methods

### PR DESCRIPTION
**Pull request by ncullen93**
I added three functions for mapping Functional Connectivity Networks from M/EEG data. This is a work-in-progress, so I would appreciate any feedback. I plan to add various FCN functions from the 7 classes of FCN measures as proposed in Wang, Benar, et al. (2014), "A sytematic framework for functional connectivity measures.":

		1) correlation
		2) h^2
		3) mutual information
		4) coherence
		5) Granger
		6) transfer entropy
		7) MVAR-frequency domain-based techniques -> A(H)

Right now, I am submitting 2 correlation functions (pearson and cosine) and one mutual information function. I include all appropriate citations. Any general discussion on how MNE would (or would not) like to incorporate functional connectivity is appreciated.

### CODE REVIEW

File: mne/connectivity/functional.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+		MI_{i,j} = \Sigma_{i,j} P(x_i,x_j) * log( P(x_i,x_j) / P(x_i)*P(x_j) )
+
+	NOTE: This function relies on '_bin_data' function, which
+	should probably be added to 'utils.py' in the 'connectivity' folder.
+	If MNE already has a data binning function, then I can use that instead.
+
+	References
+	----------
+	[1] Wange, Benar, et al. (2014), "A sytematic framework for functional
```
Comment (teonbrooks) [inline mne/connectivity/functional.py]: sytematic --> systematic

**Comment by kingjr** [comment]
Thanks for joining in! In the future, I would recommend breaking down your ENH into multiple PRs.

For this one in particular, could you make an example using mne `Epochs`? It's unclear to me how your functions can be used to iterate over trials. 

We may also want to integrate the package https://github.com/SCoT-dev/SCoT/blob/master/scot/ which already contains a bunch of connectivity measures; also see https://github.com/mne-tools/mne-python/issues/1187, and @agramfort gist https://gist.github.com/agramfort/9875439 for effective connectivity measures.

I think we should only merge the code that is ready for release, so I would recommend removing the unstable functions ('i.e. I am not getting the correct results from the hypothesis test ...'). You can develop them and ask for feedback in mne-sandbox.

You'll probably have to give a quick read to the [contribution section](http://martinos.org/mne/stable/contributing.html#contributing) to ensure your code follows the guidelines, and comply to pep8 and flake8 conventions.

You can ping me when you and Travis are ready for review.

**Comment by dengemann** [comment]
Hi Nick, thanks for sharing your code. I think this would be more appropriate for mne-sandbox. There we have more degrees for freedom. For MNE we need to spend more time with API questions, documentation and testing. With regard to contents I am a bit concerned about the API. I don't see such a clear separation between spectral and functional, in fact the former can be a subset of the latter. Ideally we should make the code fit into the spectral code that is quite elaborated and smart in handling different data inputs and is optimized for performance. As a result we might want to rename it as "functional" though. Or we would use the same private functions and expose another function to compute functional connectivity measures that don't explicitly do a spectral decomposition. I'm wondering though how relevant that would be. As I see it correlation measures and MI are often computed on spectrally decomposed inputs or filtered data. Things to be discussed.

**Comment by ncullen93** [comment]
Ok thank you for the comments everyone. I understand the code is very unpolished, but I am happy to revitalize discussion/development of more functional/effective connectivity measures. I will move this to the sandbox.. Should I just submit the same "WIP" pull request there?

Nick

**Comment by agramfort** [comment]
Should I just submit the same "WIP" pull request there?

Yes please

**Comment by ncullen93** [comment]
Ok great I have done that so I will close this. Thanks!

**Pull request closed without merging**